src/plugins/plugin.gridlines.js

```diff
@@ -0,0 +1,166 @@
+'use strict';
+
+module.exports = function(Chart) {
+	var helpers = Chart.helpers;
+	var globalDefaults = Chart.defaults.global;
+
+	// Stores all gridLines that should be displayed
+	var lines = [];
+
+	function getGridLine(chart, scale, position, tickIndex) {
+		var chartArea = chart.chartArea;
+		var scaleOptions = scale.options;
+		var gridLines = scaleOptions.gridLines;
+
+		var lineWidth, lineColor, borderDash, borderDashOffset;
+
+		if (tickIndex === (typeof scale.zeroLineIndex !== 'undefined' ? scale.zeroLineIndex : 0)) {
+			lineWidth = gridLines.zeroLineWidth;
+			lineColor = gridLines.zeroLineColor;
+			borderDash = gridLines.zeroLineBorderDash;
+			borderDashOffset = gridLines.zeroLineBorderDashOffset;
+		} else {
+			lineWidth = helpers.getValueAtIndexOrDefault(gridLines.lineWidth, tickIndex);
+			lineColor = helpers.getValueAtIndexOrDefault(gridLines.color, tickIndex);
+			borderDash = helpers.getValueOrDefault(gridLines.borderDash, globalDefaults.borderDash);
+			borderDashOffset = helpers.getValueOrDefault(gridLines.borderDashOffset, globalDefaults.borderDashOffset);
+		}
+
+		var x1, x2, y1, y2;
+
+		if (scale.isHorizontal()) {
+			x1 = x2 = position;
+			y1 = chartArea.top;
+			y2 = chartArea.bottom;
+		} else {
+			x1 = chartArea.left;
+			x2 = chartArea.right;
+			y1 = y2 = position;
+		}
+
+		return {
+			x1: x1,
+			x2: x2,
+			y1: y1,
+			y2: y2,
+			width: lineWidth,
+			color: lineColor,
+			borderDash: borderDash,
+			borderDashOffset: borderDashOffset
+		};
+	}
+
+	function collectVisibleGridLines(chart) {
+		lines = [];
+
+		// Temporary object that stores already collected gridLine positions to check and prevent gridLines from overlapping
+		var gridLinesHash = {
+			horizontal: {},
+			vertical: {}
+		};
+
+		// Marks scale border positions to prevent overlapping of gridLines and scale borders
+		helpers.each(chart.scales, function(scale) {
+			var scaleOptions = scale.options;
+
+			// gridLines.drawBorder is deprecated
+			if (scaleOptions.gridLines.drawBorder && scaleOptions.borderColor !== null && scaleOptions.borderWidth !== 0 && scaleOptions.borderWidth !== null) {
+				var glHashByOrientantion = gridLinesHash[!scale.isHorizontal() ? 'horizontal' : 'vertical'];
+				var borderPosition;
+
+				if (scale.isHorizontal()) {
+					borderPosition = scale.position === 'top' ? scale.bottom : scale.top;
+				} else {
+					borderPosition = scale.position === 'left' ? scale.right : scale.left;
+				}
+
+				glHashByOrientantion[Math.round(borderPosition)] = true;
+			}
+		});
+
+		// Collects gridLines
+		helpers.each(chart.scales, function(scale) {
+			var scaleOptions = scale.options;
+
+			if (scaleOptions.display && scaleOptions.gridLines.display && scaleOptions.gridLines.drawOnChartArea) {
+				var glHashByOrientantion = gridLinesHash[scale.isHorizontal() ? 'horizontal' : 'vertical'];
+				var position;
+
+				for (var tickIndex = 0; tickIndex < scale.ticks.length; tickIndex++) {
```

Comment:
Yes, you should try to gather vars in the same scope this way to help minification.

```javascript
helpers.each(chart.scales, function(scale) {
    var scaleOptions = scale.options;
    var glHashByOrientantion, position, i, ilen, tick;

    //...
});
```